src/test/ui/associated-types/higher-ranked-projection.bad.nll.stderr

```diff
@@ -1,8 +1,12 @@
-error: higher-ranked subtype error
+error[E0308]: mismatched types
   --> $DIR/higher-ranked-projection.rs:25:5
    |
 LL |     foo(());
-   |     ^^^^^^^
+   |     ^^^^^^^ one type is more general than the other
+   |
+   = note:   expected type `&'a ()`
+           found reference `&()`
```

Comment:
It would be nice if we could reach parity with [the non-nll output](https://github.com/rust-lang/rust/blob/056e631f89ba5f39f7742e88375737d25a4aed41/src/test/ui/associated-types/higher-ranked-projection.bad.stderr):

```
error[E0308]: mismatched types
  --> $DIR/higher-ranked-projection.rs:25:5
   |
LL |     foo(());
   |     ^^^ lifetime mismatch
   |
   = note: expected type `&'a ()`
              found type `&()`
note: the lifetime requirement is introduced here
  --> $DIR/higher-ranked-projection.rs:15:33
   |
LL |     where for<'a> &'a T: Mirror<Image=U>
   |                                 ^^^^^^^
```

Reply:
I have some good news and less good news here. The good news is, this PR uses basically the same piece of code that emits the error above, the "nice region errors placeholder errors", so it's at least not impossible to reach parity. I believe this was purposeful and one of Matthew's goals (I'm only the slightly-informed messenger here).

The "bad" news is the data: I don't think I have access to the actual same `ObligationCause`s in this case (or couldn't find them when I looked) so if you or @nikomatsakis (I know he won't see this notification, it's more of a message for when/if he also reviews this PR) can point me in the right direction to add them, I'll be glad to work on this more.

To be more specific, this error note is emitted by [`report_placeholder_failure`](https://github.com/rust-lang/rust/blob/b5a2ccee81406303324016d03399fac68ceb6718/compiler/rustc_infer/src/infer/error_reporting/note.rs#L327) which specifically looks for a `BindingObligation` cause. 

Under full NLLs (and that could be specific to this PR rather than in general, I don't know), this is an `UpperBoundUniverseConflict` but with types, so [the nice region error is not actually emitted](https://github.com/rust-lang/rust/blob/b5a2ccee81406303324016d03399fac68ceb6718/compiler/rustc_infer/src/infer/error_reporting/nice_region_error/placeholder_error.rs#L166) as these placeholder conflicts are only about placeholder and traits mismatches. The obligation available is however `MiscObligation` here. (I believe it's [this dummy one](https://github.com/rust-lang/rust/pull/86700/files#diff-29ead0842640658ccc449c0f3457618f666ba141016fe65d45e1a2d71d3cb926R217)).

Funnily enough, `report_placeholder_failure` [would emit](https://github.com/rust-lang/rust/blob/b5a2ccee81406303324016d03399fac68ceb6718/compiler/rustc_infer/src/infer/error_reporting/note.rs#L337) basically what [this PR does](https://github.com/rust-lang/rust/pull/86700/files#diff-29ead0842640658ccc449c0f3457618f666ba141016fe65d45e1a2d71d3cb926R338) in such fallback cases.

Reply:
Hmm. I'm not sure off hand how to get that obligation cause, it's probably going to take a bit of threading.

Reply:
Understood, thank you. I will look into it when I can.

Do you and @estebank agree that we could leave this topic for follow-up improvement PRs ? (Esteban has mentioned this on other suggestions, and I presume it would be ok for this suggestion as well)

Reply:
As long as we don't turn NLL on without addressing this, I'm ok with it.